File: behavioral/strategy.go
package main
import "fmt"

type ShippingStrategy interface {
    Calculate(weight float64) float64
}

type AirShipping struct{}
func (AirShipping) Calculate(w float64) float64 { return 10 + w*2 }
type SeaShipping struct{}
func (SeaShipping) Calculate(w float64) float64 { return 5 + w*0.5 }

type Order struct { strategy ShippingStrategy }
func (o *Order) SetStrategy(s ShippingStrategy) { o.strategy = s }
func (o *Order) ShippingCost(w float64) float64 { return o.strategy.Calculate(w) }

func main(){
    o := &Order{strategy: AirShipping{}}
    fmt.Println(o.ShippingCost(3.5))
    o.SetStrategy(SeaShipping{})
    fmt.Println(o.ShippingCost(3.5))
}

File: behavioral/strategy.php
<?php
interface ShippingStrategy {
    public function calculate(float $weight): float;
}

class AirShipping implements ShippingStrategy {
    public function calculate(float $weight): float { return 10 + $weight * 2; }
}
class SeaShipping implements ShippingStrategy {
    public function calculate(float $weight): float { return 5 + $weight * 0.5; }
}

class Order {
    private ShippingStrategy $strategy;
    public function __construct(ShippingStrategy $s) { $this->strategy = $s; }
    public function setStrategy(ShippingStrategy $s) { $this->strategy = $s; }
    public function shippingCost(float $weight): float { return $this->strategy->calculate($weight); }
}
$order = new Order(new AirShipping());
echo $order->shippingCost(3.5);
$order->setStrategy(new SeaShipping());
echo $order->shippingCost(3.5);
